rolling-update-kubernetes/src/main/scala/akka/rollingupdate/kubernetes/PodDeletionCostAnnotator.scala

```diff
@@ -0,0 +1,221 @@
+      log.debug("Received while on retry backoff MemberUp {}", m)
+      context.become(underRetryBackoff(deletionCost, membersByAgeDesc + m))
+
+    case ClusterEvent.MemberRemoved(m, _) =>
+      log.debug("Received while on retry backoff MemberRemoved {}", m)
+      context.become(underRetryBackoff(deletionCost, membersByAgeDesc - m))
+
+    case RetryAnnotate =>
+      updateIfNewCost(deletionCost, membersByAgeDesc)
+
+    case es => log.debug("Ignoring message {}", es)
+  }
+
+  private def costRequestToResult(newCost: Int): Future[RequestResult] = {
+    log.debug(
+      "Updating pod-deletion-cost annotation for pod: [{}] with cost: [{}]. Namespace: [{}]",
+      settings.podName,
+      newCost,
+      podNamespace
+    )
+    val request = ApiRequests.podDeletionCost(settings, apiToken, podNamespace, newCost)
+    val response =
+      clientSslContext
+        .map(http.singleRequest(request, _))
+        .getOrElse(http.singleRequest(request))
+
+    toResult(response)
+  }
+
+  private def updateIfNewCost(existingCost: Int, membersByAgeDesc: immutable.SortedSet[Member]): Unit = {
+
+    val podsToAnnotate = membersByAgeDesc.take(costSettings.annotatedPodsNr)
+    val newCost: Int = OlderCostsMore.costOf(cluster.selfMember, podsToAnnotate).getOrElse(0)
+    log.debug(
+      "Calculated cost={} (previously {}) for member={} in members by age (desc): {}",
+      newCost,
+      existingCost,
+      cluster.selfMember,
+      membersByAgeDesc)
+
+    if (newCost != existingCost) {
+      costRequestToResult(newCost).pipeTo(self)
+      context.become(onMessage(newCost, membersByAgeDesc))
+    } else context.become(onMessage(existingCost, membersByAgeDesc))
+  }
+
+}
+
+/**
+ * INTERNAL API
+ * @param annotatedPodsNr the number of members of the cluster that need to be annotated
+ * @param retryDelay fixed time delay before next attempt to annotate in case the previous one failed
+ */
+@InternalApi private[kubernetes] final case class PodDeletionCostSettings(
+    annotatedPodsNr: Int,
+    retryDelay: FiniteDuration)
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[kubernetes] object PodDeletionCostSettings {
+  val configPath: String = "pod-deletion-cost"
+  def apply(config: Config): PodDeletionCostSettings =
+    PodDeletionCostSettings(
+      config.getInt(s"$configPath.annotated-pods-nr"),
+      config.getDuration(s"$configPath.retry-delay", TimeUnit.SECONDS).seconds
+    )
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[kubernetes] object PodDeletionCostAnnotator {
+  case object RetryTimerId
+  case object RetryAnnotate
+  sealed trait RequestResult
+  case object PodAnnotated extends RequestResult
+  case class ScheduleRetry(cause: String) extends RequestResult
+  case class GiveUp(cause: String) extends RequestResult
+
+  private def toResult(futResponse: Future[HttpResponse])(implicit context: ActorContext): Future[RequestResult] = {
+    implicit val mat: Materializer = Materializer(context)
```

Comment:
This is a resource leak, the materialiser has a lifecycle and each call to `toResult` creates a new one that leaks.

Easiest is to instead take an implicit system paramameter, instead of context, that way the system materialiser is implicitly available.